rust/connlib/tunnel/src/dns.rs

```diff
@@ -592,32 +646,45 @@ mod tests {
     }
 
     #[test]
-    fn wildcard_matching() {
-        let resources = HashMap::from([("*.foo.com".to_string(), 0), ("*.com".to_string(), 1)]);
+    fn pattern_displays_without_slash() {
+        let pattern = Pattern::new("**.example.com").unwrap();
 
-        assert_eq!(match_domain(&domain("a.foo.com"), &resources), Some(0));
-        assert_eq!(match_domain(&domain("foo.com"), &resources), Some(0));
-        assert_eq!(match_domain(&domain("a.b.foo.com"), &resources), Some(0));
-        assert_eq!(match_domain(&domain("oo.com"), &resources), Some(1));
-        assert_eq!(match_domain(&domain("oo.xyz"), &resources), None);
+        assert_eq!(pattern.to_string(), "**.example.com")
     }
 
-    #[test]
-    fn question_mark_matching() {
-        let resources = HashMap::from([("?.bar.com".to_string(), 1)]);
+    #[test_case("**.example.com", "example.com"; "double star matches root domain")]
+    #[test_case("app.**.example.com", "app.bar.foo.example.com"; "double star matches multiple levels within domain")]
+    #[test_case("**.example.com", "foo.example.com"; "double star matches one level")]
+    #[test_case("**.example.com", "foo.bar.example.com"; "double star matches two levels")]
+    #[test_case("*.example.com", "foo.example.com"; "single star matches one level")]
+    #[test_case("*.example.com", "example.com"; "single star matches root domain")]
+    #[test_case("foo.*.example.com", "foo.bar.example.com"; "single star matches one domain within domain")]
+    #[test_case("app.*.*.example.com", "app.foo.bar.example.com"; "single star can appear on multiple levels")]
+    #[test_case("app.f??.example.com", "app.foo.example.com"; "question mark matches one letter")]
+    #[test_case("app.example.com", "app.example.com"; "matches literal domain")]
+    #[test_case("*?*.example.com", "app.example.com"; "mix of * and ?")]
+    #[test_case("app.**.web.**.example.com", "app.web.example.com"; "multiple double star within domain")]
+
+    fn domain_pattern_matches(pattern: &str, domain: &str) {
+        let pattern = Pattern::new(pattern).unwrap();
+        let candidate = Candidate::from_str(domain).unwrap();
 
-        assert_eq!(match_domain(&domain("a.bar.com"), &resources), Some(1));
-        assert_eq!(match_domain(&domain("bar.com"), &resources), Some(1));
-        assert_eq!(match_domain(&domain("a.b.bar.com"), &resources), None);
+        let matches = pattern.matches(&candidate);
+
+        assert!(matches);
     }
 
-    #[test]
-    fn exact_matching() {
-        let resources = HashMap::from([("baz.com".to_string(), 2)]);
+    #[test_case("app.*.example.com", "app.foo.bar.example.com"; "single star does not match two level")]
+    #[test_case("app.*com", "app.foo.com"; "single star does not match dot")]
+    // #[test_case("app.**com", "app.foo.com"; "double star does not match dot")]
```

Comment:
What about this case @jamilbk ? Currently, `**` also matched across dots which implies that you can match `.com` with it as well. Is that what we want?

Reply:
Perhaps we should require it to end with _something_. 

`**com` should match I would think yeah.

Reply:
> `**com` should match I would think yeah.

Meaning I can delete this testcase and we are okay with the behaviour that `app.**com` matches `app.com` or `app.foo.com`? The more I think about it, the more natural it is. `**` matches across dots so why wouldn't it also match one at the boundaries.

Reply:
```suggestion
```